src/consumer/consumer.js

```diff
@@ -44,43 +51,105 @@ function consumerPromise({id, init}) {
+
 			if (typeof msg === 'undefined' || !msg) {
 				log.error('Empty Message received. Skipping.');
 				return;
 			}
 
+			// Attempts left for this message
+			let attemptsLeft = retryLimit;
 			const record = JSON.parse(msg.content.toString());
-			const error = await consumeRecord({workerId: id, ...record});
-
-			switch (error) {
-				case Error.NONE:
-					log.info(
-						`[CONSUMER::${id}] Read message successfully
-						\r${record}`
-					);
-					queue.acknowledge(msg);
-					break;
-				case Error.INVALID_RECORD:
-				case Error.RECORD_ENTITY_NOT_FOUND:
-					log.warning(
-						`[CONSUMER::${id}] ${error} - \
-						\r Skipping the errored record.`
-					);
-					queue.acknowledge(msg);
-					break;
-				case Error.TRANSACTION_ERROR:
-					log.warning(
-						`[CONSUMER::${id}] ${error} Setting up for reinsertion.
-						\r Record for reference:: \n ${record}`
-					);
-					break;
-				default: break;
-			}
-		}
 
+			// Manages consume record retries
+			async.whilst(
+				// Test whose return value if is false, stops the iteration
+				// When number of attempts finish up, the test should return F
+				() => attemptsLeft > 0,
+
+				// Function repeated upon transaction error for retries times
+				// Manages async record consumption
+				async (cb) => {
+					if (attemptsLeft < retryLimit) {
+						log.info(`
+						\r ##################################################
+						\r Restarting import process....
+						\r ##################################################
+						`);
+					}
+
+					log.info('Running async function');
+					const error = await consumeRecord({
+						importRecord,
+						workerId: id,
+						...record
+					});
+
+					switch (error) {
+						case Error.NONE:
+							// On success, we don't need to retry again
+							attemptsLeft = 0;
+							log.info(
+								`[CONSUMER::${id}] Read message successfully
+								\r${record}`
+							);
+							queue.acknowledge(msg);
+							break;
+
+						case Error.INVALID_RECORD:
+						case Error.RECORD_ENTITY_NOT_FOUND:
+							// In case of invalid records, we don't try again
+							attemptsLeft = 0;
+							log.warning(
+								`[CONSUMER::${id}] ${error} -\
+								\r Skipping the errored record.`
+							);
+							// As we're not retrying, we acknowledge the message
+							queue.acknowledge(msg);
+							break;
+
+						case Error.TRANSACTION_ERROR:
+							// In case of transaction errors, we retry a number
+							// 		of times before giving up
+							attemptsLeft--;
+
+							// Issue a warning in case of transaction error
+							log.warning(
+								`[CONSUMER::${id}] ${error} Setting up for\
+								\r reinsertion. Record for reference:
+								\r ${msg}
+								\r Attempts left: ${attemptsLeft}`
+							);
+
+							// If no more attempts left, acknowledge the message
+							if (!attemptsLeft) {
+								log.info('No more attempts left.',
+									'Acknowledging the message.');
+								queue.acknowledge(msg);
```

Comment:
Right now, I acknowledge the message if it fails multiple times - under the assumption that either something is wrong with the message. It makes sense that it might be that something breaks on the consumer end - postgres server shuts down or some bug effects the process. In that case we would be better off sending `nack` to the queue. But if the message itself is corrupt, then it will linger the queue and effect the import process process.
One way to deal with it could be add the presently missing check where we first check if the import_id and timestamp on the message is consistent with desired requirements for the import to happen (using link_imports table). That way we can run producers as many times we want on the dumps.

Reply:
I feel keeping a separate queue to catch erred records would be a good idea, especially as we might want to check what errors occurred and replay them later.